Add support for --sudo to run docker with sudo on host

lloesche opened the pull request:
This PR adds
```
$ sudo dcos tunnel vpn --user=centos --docker='sudo docker'
```
to support our default AWS CentOS7 clusters where the `centos` user doesn't have access to the `docker` daemon.

Plus if `docker version` fails it'll try to run `sudo docker version` instead.

It also adds more error checking to make sure the commands we run on the master execute successfully.

nlsun commented:
I was actually thinking about letting the user define their own docker command. So they can pass in `sudo docker`, for future proofing purposes (in case someone has to do `sudo -u dockeruser docker` for whatever reason).

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -519,9 +523,10 @@ def gen_hosts(ssh_client):
-    scom = 'docker exec {} bash -c "cat {}"'
-    scom = scom.format(container_name, remote_filepath)
+def container_cp(ssh_client, container_name, remote_filepath, local_file,
+                 docker_cmd='docker'):
```
I don't think docker_cmd should have a default here.

lloesche replied:
My thinking here was, however unlikely, someone might use your library in their code and call this method, so I shouldn't break backwards compatibility.

lloesche replied:
What about _container_cp() or __container_cp() to mark or make it private?

lloesche commented:
I like your idea to give the ability to provide a custom docker command. Though I would also like to make it easy for the user. What's currently missing completely is noticing if any of the ssh commands fail. I haven't looked into it but we need the return code of the commands executed and then handle failures. Then I was thinking to be more user friendly, instead of making it complicated for them we could just try to `sudo docker` if `docker` on it's own fails.

nlsun commented:
Ah yes, there was a note I wrote with myself that I didn't share here. I was going to add an additional check in the beginning to see if the docker command works (like a `<dockercommand> ps` and check exit code or something). You can also do that in this patch if you'd like.

nlsun commented:
Trying `docker` and then `sudo docker` seems fine.

So it sounds like there is:
if <dockercommand> set, then try `<dockercommand> ps`, error out if it doesn't work.
else `docker ps` > `sudo docker ps` error out if it doesn't work and suggest setting <dockercommand> while mentioning which commands were tried.

lloesche commented:
Ok, I removed `--sudo` and replaced it with `--docker='sudo docker'`. Also added lots of error checking and the automatic try `docker` then try `sudo docker`.

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -182,7 +186,7 @@ def _info():
 def ssh_exec_wait(output):
     _, stdout, _ = output
-    stdout.channel.recv_exit_status()
+    return True if stdout.channel.recv_exit_status() == 0 else False
```
You can just `return stdout.channel.recv_exit_status() == 0`

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -534,7 +542,7 @@ def run_vpn(command, output_file):
 
 def _vpn(port, config_file, user, privileged, ssh_port, host,
-         openvpn_container, vpn_client):
+         openvpn_container, vpn_client, docker_cmd='docker'):
```
docopt already fills these out for you. Also, I think it's actually useful to receive a None here. See later comment.

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -611,11 +621,23 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
+        _, query_stdout, _ = client.exec_command(scom, get_pty=True)
+        if query_stdout.channel.recv_exit_status() != 0\
+                and not docker_cmd.startswith('sudo'):
+            docker_cmd = 'sudo ' + docker_cmd
```
I don't think we should be appending `sudo` if the user specified a command. We should only automatically try `sudo` if the user did not specify any flag, in which case docopt will set the value to `None`.

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -611,11 +621,23 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
+                and not docker_cmd.startswith('sudo'):
+            docker_cmd = 'sudo ' + docker_cmd
+            scom = '{} version'.format(docker_cmd)
+            _, query_stdout, _ = client.exec_command(scom, get_pty=True)
```
Since it seems like this this `client.exec_command()` and subsequent check for `recv_exit_status()` and raising exception is done everywhere. Maybe pull this out into a separate function?

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -611,11 +621,23 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
+            scom = '{} version'.format(docker_cmd)
+            _, query_stdout, _ = client.exec_command(scom, get_pty=True)
+            if query_stdout.channel.recv_exit_status() != 0:
+                raise Exception('Failed to execute', scom)
```
Change all the `Exception` to `DCOSException`

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -632,37 +654,49 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
-        ssh_exec_wait(server_setup)
 
-        scom = ("""docker inspect --format='"""
+        if not vpn_server.is_alive():
```
This check doesn't catch anything that `ssh_exec_wait()` does not.

nlsun replied:
The hanging behavior you observed, where the vpn server never came up and the wait message was hanging forever, was because of the `until` loop inside the `scom`, not because the thread terminated. The case where the container never comes up because the docker command was invalid is caught by the new checking logic so it won't happen again.

The only way to prevent the hanging here is to introduce a timeout. But I'm not sure if that would be helpful here, as the user can just terminate the command on their own.

lloesche replied:
I thought about that too but figured there could be many reasons why the server container dies, unrelated to the docker command. Maybe someone accidentally breaks the Dockerfile and creates a container that can't be started to begin with. Or maybe there's a missconfiguration on the host that allows you to execute `docker version` but you can't actually start any containers. For those cases I figured it makes sense to check if the container is still running before we run another ssh command. After all it's a very inexpensive check compared to the cost of a ssh connection.

nlsun replied:
The `until` loop will hang because the server container never started in the first place. It's not the cost, but rather that regardless of the reason, if the container never comes up, or if the `docker` command can't run, the `until` loop will never terminate and we will never even encounter future checks.

If it gets past the `until` loop then the vpn server successfully started at some point in time. In that case, if it crashes, we will also exit here because the vpn client will fail to connect which is the intended behavior.

lloesche replied:
Sure, but what I'm getting at is that if the VPN server crashes early on for some reason than with this check we never get stuck in the until loop and it's virtually free.

lloesche commented on dcos_tunnel/cli.py:
```diff
@@ -75,6 +76,8 @@
         Manually specify which host to connect to (it will override attempts to
         detect which host to connect to)
+    --docker=<docker-command>
+        Docker client to run [default: docker]
```
We're already setting a hard coded default right now. This just maintains that default.

lloesche commented on dcos_tunnel/cli.py:
```diff
@@ -33,6 +33,7 @@
                     [--host=<host>]
                     [--container=<container>]
                     [--client=<path>]
+                    [--docker=<docker-command>]
```
What do you suggest? `--docker-remote-command` seemed very long. Maybe `--docker-client` since we're already using `--client` to refer to the openvpn binary? Maybe we should change client to `--openvpn-client` to make it consistent.

nlsun replied:
i think `--remote-docker` seems reasonable.

lloesche replied:
Ok I'll make it `--remote-docker` for now though on first read I would assume a hostname or IP as arg to anything --remote*.

nlsun replied:
Hmm we can change the name, I'm not particularly attached to `remote-docker`.

lloesche commented:
Ok, implemented all suggested improvements and refactored exec/raise parts into a function. Also fixed some DCOS -> DC/OS wording. Let me know if this can land or if there's anything left to change. Thanks!

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -555,6 +562,8 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
     :param vpn_client: Relative or absolute path to openvpn client
     :type vpn_client: str
+    :param docker_cmd: The docker client command
+    :type docker_cmd: str
```
str | None

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -611,11 +620,26 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
         clientfile, clientpath = client_tup
 
+        try_sudo = False
+        if not docker_cmd:
```
`if docker_cmd is None`, otherwise this will also match the empty string

nlsun commented on dcos_tunnel/cli.py:
```diff
@@ -611,11 +620,26 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
+                scom = '{} version'.format(docker_cmd)
+                ssh_exec_fatal(client, scom)
+            else:
+                raise
```
Maybe raise a more informative "tried `docker` and `sudo docker` but it didn't work, set a docker command with something that works" here.

nlsun replied:
Ah actually this one is correct, there should be a try/except around the `ssh_exec_fatal` in the `try_sudo` "if statement" instead that has the more informative error message.

lloesche commented:
Ok I added the ability to give additional hints to the user when a remote command fails. Was debating whether I should add it in addition to the default error message or in it's place. Right now it's shown in addition.

nlsun commented:
Looks good, if you're also satisfied then I'll merge it.

lloesche commented:
:shipit:
